readme.md
# Firebase Functions Deploy Helper (FFDH)

A simple Node CLI that simplifies the process of deploying Firebase Functions for large projects.

As I started work on my first Firebase project, I quickly had a catalog of 57 functions I regularly deployed. Seemingly depending on how many functions I'd modified since my least deploy or on the time of day, I regularly received deployment errors and the `firebase deploy` command rarely deployed all of my functions. In seeking help from Google, I learned that there are [quota limits](https://firebase.google.com/docs/functions/quotas#quota_limits_for_firebase_cli_deployment) when deploying functions:

For each function that the Firebase CLI deploys, these types of rate and time limits are affected:

- API calls (READ) - 1 call per deployment, no matter how many functions
  - Limit: 5000 per 100 seconds
- API calls (WRITE) - 1 call per function
  - Limit: 80 per 100 seconds

I'm not sure why my little 57-function project broke the quotas, but apparently they did.

Google's recommendation is to deploy the functions in batches, but since I was hacking around in my functions it was rather difficult to keep track of what I wanted to deploy and a pain to type the long batch deploy command:

```shell
firebase deploy --only functions:function1,functions:function2,functions:function3,functions:functionX
```

I found that if I did a full deploy multiple times, it would randomly fail on a different set of functions each time and I could deploy everything (usually) with two of these:

```shell
firebase deploy --only functions
```

Not the best solution, but it worked (most of the time); it just took too long for me.

After a while, it occurred to me that I could automate some of this. My functions catalog is organized around my backend application's record types (companies, contacts, campaigns, etc.) so I thought first about using command files (`.cmd`, `.bat`, `.sh`) for the different groups, but that would be a pain to implement and maintain. I didn't want 10 `.cmd` and 10 `.sh` scripts hanging around in my project. Also, sometimes I'd modify all of the project's `get` functions, so that approach wouldn't work there.

I decided to build a simple CLI that enabled me to publish a subset of my functions - and here we are.

The CLI command delivered here is `ffdh` (Firebase Functions Deployment Helper); it allows you to:

- Deploy functions in batches. For example, you can define 5 batches, then issue separate commands to deploy each one of the batches. If you wait enough time between deploys (100 seconds as shown in the content above) then you can get around the Firebase quota limitations.
- Deploy functions based on the start and/or the end of function names in your project

## Installation

To install the module, open a terminal window or command prompt and execute the following command:

```shell
npm install -g firebase-functions-deploy-helper
```

This installs a `ffdh` command in the global scope.

## Initialization

To operate, the `ffdh` command requires access to the list of functions in your project. I know I could parse the imports in the project's `functions/index.js` or `functions/index.ts` file to build that list, but since Google recommends several ways to [organize multiple functions](https://firebase.google.com/docs/functions/organize-functions), I knew I couldn't accurately do this for all projects. Instead, I decided to use a simple JSON array maintained by the developer (that's you) for my list of function names.

To initialize your project to use this module, create a file called `ffdh.json` in your Firebase project folder (the project root, not the `functions` folder). Populate the file with a JSON array of function names as shown in the following example:

```typescript
[
    "companyDelete",
    "companyFind",
    "companyGet",
    "companyList",
    "companyUpdate",
    "contactDelete",
    "contactFind",
    "contactGet",
    "contactList",
    "contactUpdate",
    "dashGetStats"
]
```

**Note:** It doesn't matter if the list is sorted, but I sorted mine just because.

Unfortunately, you'll have to manually manage this file's content as you add and delete functions from your project. If you want to propose an accurate (and easy to maintain) mechanism to do this automatically, please submit a pull request.

## Usage

The `ffdh` command supports the following command-line options:

* `-V`, `--version`: Display the module's version number
* `-s`, `--start`: Search the start of function name for a specific string
* `-e`, `--end`: Search end of function name for a specific string
* `-b`, `--batches`: Deploy a batch of functions of a specific size
* `-bn`, `--batch`: The batch number to deploy
* `-d`, `--debug`: Enable debug mode which displays additional information to the console during processing
* `-h`, `--help`: Displays help content

The following sections describe how to use them.

### Deploying Functions in Batches

If you have a big Functions project, the easiest way to deploy all the functions is to do them in batches.  With batch deployment, you decide how many batches you want and the program automatically manages deploying the right number of functions for each batch.

Say for example that you wanted to deploy your functions in 4 batches, open a terminal window or command prompt in your Firebase project folder then execute the following command:

```shell
ffdh -b 4
```

This tells the deployment helper that you want to deploy in 4 batches and to deploy the first batch (when you omit the batch number [`-bn`], the helper assumes you want to deploy the first batch, so it sets -bn to 1). The helper figures out how many functions to deploy in each batch, then gets to work. It doesn't do anything to deploy functions evenly across batches, so depending on the number of functions in your project, the last batch could be rather small.

You can also accomplish the same thing using the following command:

```shell
ffdh -b 4 -bn 1
```

To deploy subsequent batches, wait a little time to let Firebase reset its quota timer (which looks to be 100 seconds, if the documentation is correct) then repeat the command using a different batch number:

```shell
ffdh -b 4 -bn 2
```

To deploy the functions, the helper simply builds that knarly `firebase deploy` command shown earlier, then executes the command automatically.

### Deploying Functions via Search

Sometimes you make changes to groups of functions, either functions that perform similar functions against different tables, or functions for a particular application area. The deployment helper accommodates this use case using search. When deploying functions, you can chose to select functions for deployment using a search string applied to the start and/or end of the function name.

For example, to deploy all of the `company` functions in the project (shown in the functions list displayed earlier), you can deploy them using the following command:

```shell
ffdh -s company
```

This deploys all functions whose names begin with `company` ; in this example: the `companyDelete`, `companyFind`, `companyGet`, `companyList`, `companyUpdate` functions.

If you recently modified all of the `Get` functions in the project, then you can deploy them using the following command:

```shell
ffdh -e Get
```

This deploys all functions in the project that end with the specified string; in this example: the `companyGet` and `contactGet` functions.

**Note:** Search strings are case sensitive; if you used a search string of `get` you wouldn't get any matches since `Get` is capitalized in the function names.

You can also deploy functions using search strings for both the start and end of the function name:

```shell
ffdh -s company -e e
```

In this silly example, the helper would deploy all functions that started with `company` and ended with `e` which, for this example, happens to be `companyDelete` and `companyUpdate`.

***

You can find information on many different topics on my [personal blog](http://www.johnwargo.com). Learn about all of my publications at [John Wargo Books](http://www.johnwargobooks.com).

If you find this code useful and feel like thanking me for providing it, please consider <a href="https://www.buymeacoffee.com/johnwargo" target="_blank">Buying Me a Coffee</a>, or making a purchase from [my Amazon Wish List](https://amzn.com/w/1WI6AAUKPT5P9).

release-notes.md
# Release Notes

## 0.0.5

December 16, 2021: A user uncovered an issue with the module and in researching the solution (I'm still not sure this fixes the user's problem) I noticed that for new Firebase functions, the `firebase.json` file no longer has a `source` property for the `functions` object. So, I refactored the code so it sets the functions source code folder to a default value of `functions` if the property is not set in the file.


ffdh.js
#!/usr/bin/env node
"use strict";
var chalk = require('chalk');
var fs = require('fs');
var logger = require('cli-logger');
var path = require('path');
var program = require('commander');
var shell = require('shelljs');
var cp = require("child_process");
var packageDotJSON = require('./package.json');
var APP_NAME = '\nFirebase Functions Deploy Helper (ffdh)';
var APP_AUTHOR = 'by John M. Wargo (https://johnwargo.com)';
var CURRENT_PATH = process.cwd();
var EXIT_HEADING = chalk.red('Exiting:');
var FIREBASE_CONFIG_FILE = 'firebase.json';
var FUNCTIONS_FILE = 'ffdh.json';
var FUNCTIONS_STRING = 'functions:';
var MAX_BATCHES = 25;
var functionsList;
var log = logger();
function checkFile(filePath) {
    log.debug("Locating ".concat(filePath));
    try {
        return fs.existsSync(filePath);
    }
    catch (err) {
        log.error("checkFile error: ".concat(err));
        return false;
    }
}
function checkDirectory(filePath) {
    log.debug("Locating ".concat(filePath));
    if (fs.existsSync(filePath)) {
        try {
            var stats = fs.statSync(filePath);
            if (stats) {
                return stats.isDirectory;
            }
            else {
                return false;
            }
        }
        catch (err) {
            log.error("checkDirectory error: ".concat(err));
            return false;
        }
    }
    else {
        return false;
    }
}
function isValidConfig() {
    log.debug(chalk.yellow('\nValidating Firebase project'));
    var filePath = path.join(CURRENT_PATH, FUNCTIONS_FILE);
    if (!checkFile(filePath)) {
        log.info("".concat(EXIT_HEADING, " Unable to locate the ").concat(filePath, " file\n"));
        return false;
    }
    else {
        log.debug("Located ".concat(filePath));
        functionsList = require(filePath);
    }
    filePath = path.join(CURRENT_PATH, FIREBASE_CONFIG_FILE);
    if (!checkFile(filePath)) {
        log.info("".concat(EXIT_HEADING, " Unable to locate the ").concat(filePath, " file\n"));
        return false;
    }
    else {
        log.debug("Located ".concat(filePath));
    }
    log.debug("Loading Firebase configuration from ".concat(FIREBASE_CONFIG_FILE));
    var firebaseConfig;
    try {
        firebaseConfig = require("./".concat(FIREBASE_CONFIG_FILE));
        log.debug("Firebase project configuration loaded");
    }
    catch (err) {
        log.info("".concat(EXIT_HEADING, " Unable to load the ").concat(FIREBASE_CONFIG_FILE, " file\n"));
        return false;
    }
    var sourceStr = '';
    if (firebaseConfig.functions.source) {
        sourceStr = firebaseConfig.functions.source;
    }
    if (sourceStr.length === 0) {
        sourceStr = 'functions';
        log.debug("Functions folder location: '".concat(sourceStr, "' (default)"));
    }
    else {
        log.debug("Functions folder location: '".concat(sourceStr, "'"));
    }
    if (sourceStr) {
        filePath = path.join(CURRENT_PATH, sourceStr);
        if (!checkDirectory(filePath)) {
            log.info("".concat(EXIT_HEADING, " Unable to locate the ").concat(filePath, " folder\n"));
            return false;
        }
        else {
            log.debug("Located ".concat(filePath));
        }
    }
    else {
        log.info("".concat(EXIT_HEADING, " Unable to determine the Functions source folder\n"));
        return false;
    }
    log.debug('Looking for Firebase CLI command');
    var res = shell.which('firebase');
    if (res) {
        filePath = res.toString();
        log.debug("firebase command at ".concat(filePath));
        if (!filePath) {
            log.info(EXIT_HEADING + ' Unable to locate the Firebase command\n');
            return false;
        }
        else {
            log.debug("Firebase command found at ".concat(path.dirname(filePath)));
        }
        log.debug(chalk.green('We have a Firebase project'));
        return true;
    }
    else {
        log.info(EXIT_HEADING + ' Unable to locate the Firebase command\n');
        return false;
    }
}
function processBatch(batches, batch) {
    var resultsArray = [];
    var batchSize = Math.ceil(functionsList.length / batches);
    log.debug("Batch: ".concat(batch, " of ").concat(batches, ", size ").concat(batchSize));
    var start = batchSize * (batch - 1);
    var end = start + batchSize;
    log.debug("From ".concat(start, " to ").concat(end - 1));
    resultsArray = functionsList.slice(start, end);
    resultsArray = resultsArray.map(function (func) { return FUNCTIONS_STRING + func; });
    return resultsArray.join(',');
}
function processSearch(start, end) {
    var resultsArray = [];
    if (start && end) {
        resultsArray = functionsList.filter(function (func) {
            return func.startsWith(start) && func.endsWith(end);
        });
    }
    else {
        if (start) {
            resultsArray = functionsList.filter(function (func) {
                return func.startsWith(start);
            });
        }
        else {
            resultsArray = functionsList.filter(function (func) {
                return func.endsWith(end);
            });
        }
    }
    resultsArray = resultsArray.map(function (func) { return FUNCTIONS_STRING + func; });
    return resultsArray.join(',');
}
console.log(APP_NAME);
program.version(packageDotJSON.version);
program.option('-s, --start <searchStr>', 'Search start of function name for <string>');
program.option('-e, --end <searchStr>', 'Search end of function name for <string>');
program.option('-b, --batches <number>', 'Deploy functions in <number> batches');
program.option('-bn, --batch <number>', 'When deploying via batches, specifies which batch number');
program.option('-d, --debug', 'Output extra information during operation');
program.parse();
var options = program.opts();
if (options.debug) {
    log.level(log.DEBUG);
}
else {
    log.level(log.INFO);
}
log.debug(APP_AUTHOR);
log.debug("Version: ".concat(packageDotJSON.version));
log.debug('Command Options:', options);
if (isValidConfig()) {
    var strFunctionList = void 0;
    if (options.batches) {
        options.batch = options.batch ? options.batch : "1";
        var batches = parseInt(options.batches, 10);
        if (batches < 1 || batches > MAX_BATCHES) {
            log.info(chalk.red("Invalid iterations value: ".concat(batches, " (Must be 1-").concat(MAX_BATCHES, ")")));
            process.exit(1);
        }
        var batch = parseInt(options.batch, 10);
        if (batch < 1 || batch > batches) {
            log.info(chalk.red("Invalid iteration value: ".concat(batch, " (Must be 1-").concat(batches, ")")));
            process.exit(1);
        }
        strFunctionList = processBatch(batches, batch);
    }
    else {
        if (!options.start && !options.end) {
            log.info(chalk.red('Nothing to do here (missing actionable parameters)'));
            process.exit(1);
        }
        strFunctionList = processSearch(options.start, options.end);
    }
    if (strFunctionList.length > 0) {
        try {
            var cmd = "firebase deploy --only ".concat(strFunctionList);
            log.info(chalk.yellow('Executing:'), cmd);
            cp.execSync(cmd, { stdio: 'inherit' });
        }
        catch (e) {
            log.warn(e);
        }
    }
    else {
        log.info(chalk.red('No function match for specified options'));
        process.exit(1);
    }
}


src/ffdh.ts
#!/usr/bin/env node
/**********************************************************
 * Firebase Functions Deployment Helper
 * by John M. Wargo
 * 
 * Simplifies deployment of a subset of a Firebase
 * project's functions. I created this thing because I 
 * constantly ran up against Firebase quote limits when
 * deploying my functions
 **********************************************************/

// How to get a list of exported functions
// https://stackoverflow.com/questions/25529290/node-js-module-how-to-get-list-of-exported-functions

const chalk = require('chalk');
const fs = require('fs');
const logger = require('cli-logger');
const path = require('path');
const program = require('commander');
const shell = require('shelljs');
const cp = require("child_process");
// const { execSync } = require("child_process");

// https://stackoverflow.com/questions/9153571/is-there-a-way-to-get-version-from-package-json-in-nodejs-code
const packageDotJSON = require('./package.json');

// constants
const APP_NAME = '\nFirebase Functions Deploy Helper (ffdh)';
const APP_AUTHOR = 'by John M. Wargo (https://johnwargo.com)';
// const COMMAND_ROOT = 'firebase deploy --only ';
const CURRENT_PATH = process.cwd();
const EXIT_HEADING = chalk.red('Exiting:');
const FIREBASE_CONFIG_FILE = 'firebase.json';
const FUNCTIONS_FILE = 'ffdh.json';
const FUNCTIONS_STRING = 'functions:';
const MAX_BATCHES = 25;

var functionsList: any;
var log = logger();

function checkFile(filePath: string): boolean {
  // log.debug(`checkFile(${filePath})`);
  log.debug(`Locating ${filePath}`);
  try {
    return fs.existsSync(filePath);
  } catch (err) {
    log.error(`checkFile error: ${err}`);
    return false;
  }
}

function checkDirectory(filePath: string): boolean {
  log.debug(`Locating ${filePath}`);
  // log.debug(`checkDirectory(${filePath})`);
  // does the folder exist?
  if (fs.existsSync(filePath)) {
    // Check to see if it's a folder
    try {
      let stats = fs.statSync(filePath);
      if (stats) {
        return stats.isDirectory;
      } else {
        return false;
      }
    } catch (err) {
      log.error(`checkDirectory error: ${err}`);
      return false;
    }
  } else {
    return false;
  }
}

function isValidConfig(): Boolean {
  // Make sure this is a Flutter project
  log.debug(chalk.yellow('\nValidating Firebase project'));

  // Does the functions list exist?  
  let filePath = path.join(CURRENT_PATH, FUNCTIONS_FILE);
  if (!checkFile(filePath)) {
    log.info(`${EXIT_HEADING} Unable to locate the ${filePath} file\n`);
    return false;
  } else {
    log.debug(`Located ${filePath}`);
    functionsList = require(filePath);
  }

  // Does the config file exist?
  filePath = path.join(CURRENT_PATH, FIREBASE_CONFIG_FILE);
  if (!checkFile(filePath)) {
    log.info(`${EXIT_HEADING} Unable to locate the ${filePath} file\n`);
    return false;
  } else {
    log.debug(`Located ${filePath}`);
  }

  // load the Firebase config file
  log.debug(`Loading Firebase configuration from ${FIREBASE_CONFIG_FILE}`);
  var firebaseConfig;
  try {
    firebaseConfig = require(`./${FIREBASE_CONFIG_FILE}`);
    log.debug(`Firebase project configuration loaded`);
  } catch (err) {
    log.info(`${EXIT_HEADING} Unable to load the ${FIREBASE_CONFIG_FILE} file\n`);
    return false
  }

  /**
   * 12/16/2021 - later versions of Firebase don't populate the `source` property, 
   * so I refactored this to set a default value if it's not set.
   */
  var sourceStr: string = '';
  // do we have a `source` property in the functions object?
  if (firebaseConfig.functions.source) {
    // then use it for the functions folder location
    sourceStr = firebaseConfig.functions.source;
  }
  // Did we get a source folder?  
  if (sourceStr.length === 0) {
    // use the default folder name
    sourceStr = 'functions';
    log.debug(`Functions folder location: '${sourceStr}' (default)`);
  } else {
    log.debug(`Functions folder location: '${sourceStr}'`);
  }

  // do we have a value?  We should, but check anyway
  if (sourceStr) {
    // does the path exist?
    filePath = path.join(CURRENT_PATH, sourceStr);
    if (!checkDirectory(filePath)) {
      log.info(`${EXIT_HEADING} Unable to locate the ${filePath} folder\n`);
      return false;
    } else {
      log.debug(`Located ${filePath}`);
    }
  } else {
    log.info(`${EXIT_HEADING} Unable to determine the Functions source folder\n`);
    return false;
  }

  // is the Firebase CLI installed?
  log.debug('Looking for Firebase CLI command');
  let res = shell.which('firebase');
  if (res) {
    filePath = res.toString();
    log.debug(`firebase command at ${filePath}`);
    if (!filePath) {
      log.info(EXIT_HEADING + ' Unable to locate the Firebase command\n');
      return false;
    } else {
      log.debug(`Firebase command found at ${path.dirname(filePath)}`);
    }
    log.debug(chalk.green('We have a Firebase project'));
    return true;
  } else {
    log.info(EXIT_HEADING + ' Unable to locate the Firebase command\n');
    return false;
  }
}

function processBatch(batches: number, batch: number,): string {
  let resultsArray: string[] = [];
  let batchSize = Math.ceil(functionsList.length / batches);
  log.debug(`Batch: ${batch} of ${batches}, size ${batchSize}`);

  const start = batchSize * (batch - 1);
  const end = start + batchSize;
  log.debug(`From ${start} to ${end - 1}`);
  // grab from start to one less than end
  resultsArray = functionsList.slice(start, end);
  // Add 'functions:' to the beginning of every function name
  resultsArray = resultsArray.map(func => FUNCTIONS_STRING + func);
  // split the array into a comma-separated list
  return resultsArray.join(',');
}

function processSearch(start: string, end: string): string {
  let resultsArray: string[] = [];
  if (start && end) {
    resultsArray = functionsList.filter((func: string) => {
      return func.startsWith(start) && func.endsWith(end);
    });
  } else {
    if (start) {
      resultsArray = functionsList.filter((func: string) => {
        return func.startsWith(start);
      });
    } else {
      resultsArray = functionsList.filter((func: string) => {
        return func.endsWith(end);
      });
    }
  }
  resultsArray = resultsArray.map(func => FUNCTIONS_STRING + func);
  return resultsArray.join(',');
}

console.log(APP_NAME);
program.version(packageDotJSON.version);
program.option('-s, --start <searchStr>', 'Search start of function name for <string>');
program.option('-e, --end <searchStr>', 'Search end of function name for <string>');
program.option('-b, --batches <number>', 'Deploy functions in <number> batches');
program.option('-bn, --batch <number>', 'When deploying via batches, specifies which batch number');
program.option('-d, --debug', 'Output extra information during operation');
program.parse();
const options = program.opts();
if (options.debug) {
  log.level(log.DEBUG);
} else {
  log.level(log.INFO);
}
// write the version number to the console
log.debug(APP_AUTHOR);
log.debug(`Version: ${packageDotJSON.version}`);
// Write the command line options to the console
log.debug('Command Options:', options);

if (isValidConfig()) {
  let strFunctionList: string;

  if (options.batches) {
    // do we have -b and not -bn? then set the default batch number (1)
    options.batch = options.batch ? options.batch : "1";
    // get numeric values for our parameters
    const batches = parseInt(options.batches, 10);
    if (batches < 1 || batches > MAX_BATCHES) {
      log.info(chalk.red(`Invalid iterations value: ${batches} (Must be 1-${MAX_BATCHES})`));
      process.exit(1);
    }
    // do we have a valid iteration?
    const batch = parseInt(options.batch, 10);
    if (batch < 1 || batch > batches) {
      log.info(chalk.red(`Invalid iteration value: ${batch} (Must be 1-${batches})`));
      process.exit(1);
    }
    // We got this far, we're good to go!
    strFunctionList = processBatch(batches, batch);
  } else {
    // do we have start or end?
    if (!options.start && !options.end) {
      // nothing to do here, goodbye
      log.info(chalk.red('Nothing to do here (missing actionable parameters)'));
      process.exit(1);
    }
    strFunctionList = processSearch(options.start, options.end);
  }

  if (strFunctionList.length > 0) {
    try {
      var cmd = `firebase deploy --only ${strFunctionList}`;
      log.info(chalk.yellow('Executing:'), cmd);
      cp.execSync(cmd, { stdio: 'inherit' });
    } catch (e) {
      log.warn(e);
    }
  } else {
    log.info(chalk.red('No function match for specified options'));
    process.exit(1);
  }

}
